subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc

```diff
@@ -89,6 +89,63 @@ See the release notes for link:https://github.com/JetBrains/kotlin/releases/tag/
 
 === Deprecations
 
+==== Deprecated `Project.buildDir` is to be replaced by `Project.layout.buildDirectory`
+
+The `Project.buildDir` property is deprecated.
+It uses eager APIs and has ordering issues if the value is read in build logic and then later modified as outputs could end up in different locations.
+
+It is replaced by a `link:{javadocPath}/org/gradle/api/file/DirectoryProperty.html[DirectoryProperty]` found at `Project.layout.buildDirectory`.
+See the `link:{dslPath}/org.gradle.api.file.ProjectLayout.html[ProjectLayout]` interface for details.
+
+Note that at this stage, Gradle will not print deprecation warnings if you still use `Project.buildDir`.
+We know this is a big change and want to give time for authors of major plugin to move away from its usage first.
+
+The switch from a `File` to a `DirectoryProperty` requires adaptations in build logic.
+The main impact is that you cannot use the property inside a `String` to expand it.
+Instead, you should leverage the `dir` and `file` methods to compute the location you want.
+
+Code like the following, either in a build script or a task configuration block:
+
+=====
+[.multi-language-sample]
+======
+.build.gradle.kts
+[source,kotlin]
+----
+file("$buildDir/myOutput.txt")
+----
+======
+[.multi-language-sample]
+======
+.build.gradle
+[source,groovy]
+----
+file("$buildDir/myOutput.txt")
+----
+======
+=====
+
+should be replaced by:
+
+=====
+[.multi-language-sample]
+======
+.build.gradle.kts
+[source,kotlin]
+----
+layout.buildDirectory.file("myOutput.txt")
```

Comment:
🤔  I think this could lead to bad UX. Users would expect the recommended replacement to work as a drop-in, which is not always the case. I think we at least should specify what types these expressions have. We can do this in a comment like `layout.buildDirectory.file("myOutput.txt") // Provider<RegularFile>`

Or, maybe better with an explicit type:
```
val dir: Provider<RegularFile> = layout.buildDirectory.file("myOutput.txt")
```